<a href="https://colab.research.google.com/github/LemdjoM/Kaggle_competition/blob/main/Kaggle.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## The Lazy Librarian

We are taking part to the competition called  Kaggle Competition.

Our mission, is to build a recommender system that predicts:

➡️ Which 10 books each user is most likely to borrow next..

# Assignment

We received three base datasets in the Data tab:

interactions.csv – user–item interaction history (who borrowed what)
items.csv – metadata for each item (title, author, subjects, etc.)
sample_submission.csv – example of the correct submission format
These files form our core training data. To make our AI Librarian smarter, we are encouraged to enrich or augment the dataset using external or AI-generated sources, such as:

🧠 GenAI tools to generate summaries, keywords, or missing metadata (titles, genres, descriptions).
🌐 Public book datasets (e.g. Project Gutenberg, Open Library, Google Books API).
Just make sure to document your data sources and generation process in your README for transparency and reproducibility.




## Getting to know our data

- userId: Unique number given to each user.

- tier: Tier of the city in which the user is residing.

- gender: Categorical feature representing the gender of the user. 1 represents male and 2 represents female.

- following_rate: Number of accounts followed by the user(feature is normalized)

- followers_avg_age: Average of age groups of all the followers of the user.

- following_avg_age: Average of age groups of all the accounts followed by the user.

- max_repetitive_punc: Maximum repititive punctuations found in the bio and comments of the user.

- num_of_hashtags_per_action: Average nubmer of hashtags used by the user per comment.

- emoji_count_per_action: Average number of emojis used by the user per comment

- punctuations_per_action: Average number of punctuations used by the user per comment.

- number_of_words_per_action: Average number of words used by the user per comment.

- avgCompletion: Average watch time completion rate of the videos.

- avgTimeSpent: Average time spent by the user on a video in seconds.

- avgDuration: Average duration of the videos that the user has watched till date.

- avgComments: Average number of comments per video watched.

- creations: Total number of videos uploaded by the user.

- content_views: Total number of videos watched.

- num_of_comments: Total number of comments made by the user (normalized)

- weekends_trails_watched_per_day: Number of videos watched on weekends per day.

- weekdays_trails_watched_per_day: Number of videos watched on weekdays per day.

- slot1_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot2_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot3_trails_watched_per_day: The day is divided into 4 slots. This feature represents the average number of videos watched in this particular time slot.

- slot4_trails_watched_per_day: data

- avgt2: Average number of followers of all the accounts followed by the user.

- age : The age of the user


In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler


interactions = pd.read_csv("https://raw.githubusercontent.com/LemdjoM/Kaggle_competition/main/data/interactions.csv")
items = pd.read_csv("https://raw.githubusercontent.com/LemdjoM/Kaggle_competition/main/data/items.csv")
sample_submission = pd.read_csv("https://raw.githubusercontent.com/LemdjoM/Kaggle_competition/main/data/sample_submission.csv")

interactions.columns = ["user_id", "item_id", "timestamp"]
items.columns = ["Title", "Author", "ISBN", "Publisher", "Subjects", "item_id"]

interactions.head(), items.head(), sample_submission.head()

(   user_id  item_id     timestamp
 0     3146     1024  1.712062e+09
 1     6145    14867  1.704817e+09
 2     2091    10068  1.686321e+09
 3     4200    10867  1.715610e+09
 4     5914     8164  1.692362e+09,
                                                Title  \
 0  Classification décimale universelle : édition ...   
 1  Les interactions dans l'enseignement des langu...   
 2  Histoire de vie et recherche biographique : pe...   
 3  Ce livre devrait me permettre de résoudre le c...   
 4                    Les années glorieuses : roman /   
 
                      Author  \
 0                       NaN   
 1  Cicurel, Francine, 1947-   
 2                       NaN   
 3     Mazas, Sylvain, 1980-   
 4   Lemaitre, Pierre, 1951-   
 
                                                 ISBN     Publisher  \
 0                          9782871303336; 2871303339   Ed du CEFAL   
 1                          9782278058327; 2278058320        Didier   
 2                          2343190194

In [ ]:
# Set the random seed for reproducibility
np.random.seed(42)

### Training a model
So it contains:

u: user_id

i: item_id

t: timestamp (UNIX)

➡️ This is exactly what we need to build all recommenders.

You can use the features `'tier', 'gender', 'following_rate', 'followers_avg_age', 'following_avg_age', 'max_repetitive_punc', 'num_of_hashtags_per_action', 'emoji_count_per_action', 'punctuations_per_action', 'number_of_words_per_action', 'avgCompletion', 'avgTimeSpent', 'avgDuration', 'avgComments', 'creations', 'content_views', 'num_of_comments', 'weekends_trails_watched_per_day', 'weekdays_trails_watched_per_day', 'slot1_trails_watched_per_day', 'slot2_trails_watched_per_day', 'slot3_trails_watched_per_day', 'slot4_trails_watched_per_day', 'avgt2'`


In [ ]:
#Your code here
features = [
    'tier', 'gender', 'following_rate', 'followers_avg_age', 'following_avg_age',
    'max_repetitive_punc', 'num_of_hashtags_per_action', 'emoji_count_per_action',
    'punctuations_per_action', 'number_of_words_per_action', 'avgCompletion',
    'avgTimeSpent', 'avgDuration', 'avgComments', 'creations', 'content_views',
    'num_of_comments', 'weekends_trails_watched_per_day', 'weekdays_trails_watched_per_day',
    'slot1_trails_watched_per_day', 'slot2_trails_watched_per_day',
    'slot3_trails_watched_per_day', 'slot4_trails_watched_per_day', 'avgt2'
]
X=train_dataset[features]
y=train_dataset[['age']]
print(f"X shape: {X.shape}")
print(f"y shape: {y.shape}")

X shape: (486049, 24)
y shape: (486049, 1)


Now do the train-test split, using 80% of the data as training and 20% as test. Set the random state to 42.

In [ ]:
#Your code here
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, shuffle=True)


Next, rescale the features in training and test data using `MinMaxScaler`.

In [ ]:
# Use the MinMaxScaler to rescale your the features in X_train and X_test
# Define the scaler
scaler = MinMaxScaler()

# Fit the scaler
scaler.fit(X_train)
# Transform the train and the test set
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

# These two steps can be merged into one (only for the train set)
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## Linear regression
Now, create a linear regression model. Train it on the training data and use the model to make predictions on the test data. How is the performance of your model on the test data?

*Note: For the liner regression model, use the default attribute value as seen in the Lab.*

In [ ]:
# do the right imports
from sklearn.linear_model import LinearRegression

# create the model
# 1. Set up the model
model = LinearRegression()
# Fit the model

# 2. Use fit
model.fit(X_train, y_train)
# 3. Check the score/accuracy
print("R^2 Score of the model: ", round(model.score(X_test, y_test), 3))
# 4. Print the coefficients of the linear model
print("Intercept: ", model.intercept_)
print("Features coefficients (weigths): ", model.coef_)# Get the coefficients, w


R^2 Score of the model:  0.663
Intercept:  [36.65759173]
Features coefficients (weigths):  [[-3.92932255e+00 -2.75015640e+00 -5.46070970e+01  1.29525341e+01
   1.64816454e+01  5.49586548e+01 -1.88236068e+01 -3.00775482e+01
  -4.74259835e+01  9.63225871e+01  1.09635537e+01  4.85035283e-01
   1.55800963e+02 -3.36215748e+01 -6.34738004e+00  2.43326582e+01
  -4.73736824e+00  2.65031459e+03  3.01403857e+03 -1.71955423e+03
  -2.48016341e+03 -2.61667125e+03 -2.64978570e+03  7.08403365e+01]]


and we do the predictions on the test set:

In [ ]:
#Your code here
# For the test data
predictions = model.predict(X_test)


What are the model's $MAE$, $MSE$ and $R^2$ ?

In [ ]:
#Your code here
from sklearn.metrics import mean_absolute_error, r2_score, mean_squared_error

print("MAE %.2f" % mean_absolute_error(y_test, predictions))
print("MSE %.2f" % mean_squared_error(y_test, predictions))
print("R^2 %.2f" % r2_score(y_test, predictions))

MAE 5.33
MSE 48.95
R^2 0.66


**Q1. What is the mean absolute error (MAE) of the model over the test set?**

*Hint : The correct value should be between 5.20 and 5.50*

Round to 2 digits (e. g. 4.32)

**Q2. What is the mean squared error (MSE) of the model over the test set?"**

*Hint : The correct value should be between 48 and 50*

Round to 2 digit (e.g. 4.32)

**Q3. What is the $R^2$ score of the model over the test set?**

*Hint : The correct value should be between 0.55 and 0.75*

Round to 2 digits (e. g. 4.32)

## Retraining on the full dataset
Training on a split of the training dataset, as you just did, is mainly for evaluating the performance of the resulting model on the unseen data and also optimizing various hyper-parameters of the model.

We can now retrain the model on the full training data prior to making predictions.

Before predicting the age of the suspects, retrain the model on the full training dataset.

You can reuse the same features as before, but make sure those are also available on the suspects dataset.

In [ ]:
#Your code here
# Suspects'available Features
suspects_features_available = set(suspects.columns)
required_features = set(features)
missing_features = required_features - suspects_features_available
extra_features = suspects_features_available - required_features

print(f"Required Features: {len(required_features)}")
print(f"Suspects's available Features: {len(suspects_features_available)}")
print(f"Missing Features: {missing_features}")
print(f"Extra Features: {extra_features}")

if missing_features:
    print("ATTENTION: Missing Features for the suspects!")
else:
    print("All the required features are available")

X_suspects = suspects[features]


Required Features: 24
Suspects's available Features: 26
Missing Features: set()
Extra Features: {'userId', 'Unnamed: 0'}
All the required features are available


In [ ]:
# Use a new MinMaxScaller to rescale your the features in X and in the suspect dataset
scaler2 = MinMaxScaler()

# Fit the scaler
scaler2.fit(X_suspects)
# Transform the train set
X_train_scaled = scaler.fit_transform(X_train)
X_suspects_scaled = scaler2.transform(X_suspects)


In [ ]:
# create the new model
# 1. Set up the model
new_model = LinearRegression()
new_model.fit(X_train_scaled, y_train)
print(f"Score R² (trained): {new_model.score(X_train_scaled, y_train):.3f}")

Score R² (trained): 0.661


In [ ]:
# Make the predictions
# For the test data
suspects_predictions = new_model.predict(X_suspects_scaled)
suspects['predicted_age'] = suspects_predictions
suspects['predicted_age'] = suspects['predicted_age'].round(1)

**Keep only the suspects between 25 and 35 years old. Use the `age` values you predicted in the previous step.**


In [ ]:
#Your code here

suspects_25_35 = suspects[
    (suspects['predicted_age'] >= 25) &
    (suspects['predicted_age'] <= 35)
].copy()

print(f"All Suspects: {suspects.shape[0]}")
print(f"25-35 year old Suspects: {suspects_25_35.shape[0]}")


# List of userIDs to check
user_ids_to_check = [255830, 123699, 410319, 785994, 757865]

print(f"UserIDs to check: {user_ids_to_check}")
print(f"Total suspects 25-35 years old: {len(suspects_25_35)}")

# Check all the userID
remaining_suspects = []
for user_id in user_ids_to_check:
    if user_id in suspects_25_35['userId'].values:
        remaining_suspects.append(user_id)
        # Récupérer l'âge prédit pour information
        age = suspects_25_35[suspects_25_35['userId'] == user_id]['predicted_age'].iloc[0]
        print(f"✅ {user_id}: In the list (predited: {age} years old)")
    else:
        print(f"❌ {user_id}: Not in the list")

print(f"\n📊 RESULT:")
print(f"Remaining Suspects: {len(remaining_suspects)}/{len(user_ids_to_check)}")



All Suspects: 729
25-35 year old Suspects: 96
UserIDs to check: [255830, 123699, 410319, 785994, 757865]
Total suspects 25-35 years old: 96
✅ 255830: In the list (predited: 34.8 years old)
✅ 123699: In the list (predited: 34.8 years old)
✅ 410319: In the list (predited: 34.8 years old)
✅ 785994: In the list (predited: 34.8 years old)
❌ 757865: Not in the list

📊 RÉSULT:
Remaining Suspects: 4/5


**Q4. Which of these users remain suspects after filtering for those aged between 25 and 35?** (check all that apply)

The SIMF is considering narrowing down the suspect pool by focusing on gender. As a preliminary step, they need to know the percentage of female suspects within the age range of 25 to 35 years old.

Calculate and report the percentage of female suspects among those aged 25 to 35 years old from the suspects you identified previously.

*Hint: In the provided dataset, 'gender' is coded with 1 representing male users and 2 representing female users.*

In [ ]:
#Your code here
# 3. Compter les femmes (gender = 2)
female_suspects = suspects_25_35[suspects_25_35['gender'] == 2]
female_count = female_suspects.shape[0]

# 4. Compter les hommes (gender = 1)
male_suspects = suspects_25_35[suspects_25_35['gender'] == 1]
male_count = male_suspects.shape[0]

total_25_35 = suspects_25_35.shape[0]
female_percentage = (female_count / total_25_35) * 100
male_percentage = (male_count / total_25_35) * 100

print(f"• Women (gender = 2): {female_count} suspects ({female_percentage:.1f}%)")
print(f"• Men (gender = 1): {male_count} suspects ({male_percentage:.1f}%)")

• Women (gender = 2): 39 suspects (40.6%)
• Men (gender = 1): 57 suspects (59.4%)


**Q5. What is the percentage of female users among the predicted suspects aged between 25 and 35 years old?**

Round to 2 digits (e. g. 4.32)



**What is the predicted age of a Female user with the following characteristics:**
* tier = 2
* following_rate = 2
* followers_avg_age = 27
* following_avg_age = 22
* max_repetitive_punc = 0.06
* num_of_hashtags_per_action = 0.8
* emoji_count_per_action = 0.3
* punctuations_per_action = 0.03
* number_of_words_per_action = 7
* avgCompletion = 12
* avgTimeSpent = 0.13
* avgDuration = 0.8
* avgComments = 0.04
* likeToDislikeRation = 0.76
* creations = 10
* content_views = 65
* num_of_comments = 12
* weekends_trails_watched_per_day = 0
* weekdays_trails_watched_per_day = 0.00353
* slot1_trails_watched_per_day = 0.032
* slot2_trails_watched_per_day = 0
* slot3_trails_watched_per_day = 0.004
* slot4_trails_watched_per_day = 2
* avgt2 = 39

*hint: make sure the features and their order corresponds to the ones used for the training.*

*hint 2: don't forget to reprocess the data as you did before*

In [ ]:
#Your code here
userDetails = [2,2,1,27,22,0.09,0.8,0.3,0.03,7,12,0.13,0.8,0.04,10,62,12,0,0.04353,0.032,0,0.004,2,39]

# 1. Create the DataFrame with given caracteristics
user_data = {
    'tier': [2],
    'gender': [2],  # 2 = woman
    'following_rate': [1],
    'followers_avg_age': [27],
    'following_avg_age': [22],
    'max_repetitive_punc': [0.09],
    'num_of_hashtags_per_action': [0.8],
    'emoji_count_per_action': [0.3],
    'punctuations_per_action': [0.03],
    'number_of_words_per_action': [7],
    'avgCompletion': [12],
    'avgTimeSpent': [0.13],
    'avgDuration': [0.8],
    'avgComments': [0.04],
    'creations': [10],
    'content_views': [62],
    'num_of_comments': [12],
    'weekends_trails_watched_per_day': [0],
    'weekdays_trails_watched_per_day': [0.04353],
    'slot1_trails_watched_per_day': [0.032],
    'slot2_trails_watched_per_day': [0],
    'slot3_trails_watched_per_day': [0.004],
    'slot4_trails_watched_per_day': [2],
    'avgt2': [39]
}

# Create the correct DataFrame with the given caracteristics
user_data_correct = {feature: [user_data.get(feature, [0])[0]] for feature in features}
user_df_correct = pd.DataFrame(user_data_correct)

# Scaling
X_user_scaled = scaler2.transform(user_df_correct)

# Prediction with the model
predicted_age_array = new_model.predict(X_user_scaled)
predicted_age = predicted_age_array.item()  # to resolve warning


# Result
print(f"PREDITED AGE: {predicted_age:.1f} ans")

PREDITED AGE: -3442.3 ans


**Q6. What is the predicted age of a female user with those characteristics using the regression model that you created?**

Enter the rounded value as an integer (e. g. 21 if your result is 21.34 or 20.79)

## Your investigation is progressing effectively, and the list of suspects is narrowing down.

**Don't forget to answer the quiz and submit your code on Moodle before the end of the deadline.**